ceph.go

```diff
@@ -605,6 +608,136 @@ func pgQuery(pgid string) *pgQueryOut {
 	return &out
 }
 
+func crushCmp(fp string) ([]pgMapping, error) {
+	out, err := runCrushCmp(fp)
+	if err != nil {
+		panic(err)
+	}
+
+	puis, err := parseCrushDiff(out)
+	if err != nil {
+		panic(err)
+	}
+
+	mappings := make([]pgMapping, 0, len(puis))
+	for _, pui := range puis {
+		for _, m := range pui.Mappings {
+			if m.From == m.To {
+				continue
+			}
+
+			mappings = append(mappings, pgMapping{
+				PgID: pui.PgID,
+				Mapping: mapping{
+					From: m.From,
+					To:   m.To,
+				},
+			})
+		}
+	}
+
+	return mappings, nil
+}
+
+func parseCrushDiff(in string) ([]*pgUpmapItem, error) {
+	var (
+		sc     = bufio.NewScanner(strings.NewReader(in))
+		puiMap = make(map[string]*pgUpmapItem)
+	)
+
+	for sc.Scan() {
+		line := strings.TrimSpace(sc.Text())
+
+		// For PGs that are part of a 3x replicated pool, each
+		// line mapping should look something like follows:
+		//
+		//  1.0	[3, 7, 8] -> [3, 7, 2]
+		if !strings.Contains(line, "->") {
+			continue
+		}
+
+		pui, err := parsePGRemapEntry(line)
+		if err != nil {
+			return nil, errors.Wrap(err, "could not parse PG mapping entry")
+		}
+
+		puiMap[pui.PgID] = pui
+	}
+
+	if err := sc.Err(); err != nil {
+		return nil, errors.Wrap(err, "failed scanning crushmap")
+	}
+
+	puis := make([]*pgUpmapItem, 0, len(puiMap))
+	for _, pui := range puiMap {
+		puis = append(puis, pui)
+	}
+
+	sort.Slice(puis, func(i, j int) bool { return puis[i].PgID < puis[j].PgID })
+	return puis, nil
+}
+
+func parsePGRemapEntry(entry string) (*pgUpmapItem, error) {
+	parts := strings.Fields(entry)
+
+	// We expect at minimum the PG mapping entry should contain 4
+	// separate sections. For e.g.:
+	//
+	// 1.0 [3] -> [3]
+	if len(parts) < 4 {
+		return nil, fmt.Errorf("incomplete PG mapping entry: %q", entry)
+	}
+
+	// There should also be even number of elements in the remapping entry.
+	if len(parts)%2 != 0 {
+		return nil, fmt.Errorf("invalid PG mapping entry: %q", entry)
+	}
+
+	var (
+		pgID   = parts[0]
+		eM, nM = make([]int, 0, len(parts)/2), make([]int, 0, len(parts)/2)
+		nMM    bool
+	)
+	for _, part := range parts[1:] {
+		if part == "->" {
+			nMM = true
+			continue
+		}
+
+		part = strings.Trim(part, "[],")
+
+		id, err := strconv.Atoi(part)
+		if err != nil {
+			return nil, errors.Wrapf(err, "could not parse pg mapping entry for %q", entry)
+		}
+
+		if nMM {
+			nM = append(nM, id)
+			continue
+		}
+
+		eM = append(eM, id)
+	}
+
+	// OSD sets should have the same length for both existing and new ones.
+	if len(eM) != len(nM) {
+		return nil, fmt.Errorf("unequal count between existing and new OSD sets within mapping: %q", entry)
+	}
+
+	enM := make([]mapping, 0, len(nM))
+	for i := range nM {
+		enM = append(enM, mapping{
+			From: eM[i],
+			To:   nM[i],
+		})
+	}
```

Comment:
So, this is where it gets tricky because whether this change is significant matters between replicated and EC. In replicated, if OSD X is in both eM and nM, it should be excluded from consideration for creating upmap entries; order doesn't matter. (Or does this not happen in the output of this tool?)

Also, is it guaranteed that `eM[i] != nM [i]` here?

Reply:
Not here, but [those are skipped here](https://github.com/digitalocean/pgremapper/pull/36/files/7a29a516a0b91693ea6d3481277b9d76b40dbbd8#diff-78d8fd876b1a5f53f8480476a217eb5912430e52185a13b409d864637250893cR625-R627) just before it is sent out.

Reply:
I have also [have a test added](https://github.com/digitalocean/pgremapper/pull/36/files#diff-97adea65ffc88737e0135e24cf79fa0e31f8a1ae5bb2c19104a9971912264dd7R42) that confirms that if it helps.

Reply:
Ah, yeah, I had added a comment on that section of code yesterday because it confused me, but it appears that comment got lost (my fault - badly timed refresh), and I understand how this fits together better now. :+1: 

We chatted in Slack and the conclusion was that you don't think the crushdiff tool will result in re-ordered replicated PGs, which satisfies my other concern here.